homeassistant/components/livisi/switch.py

```diff
@@ -23,54 +30,53 @@ async def async_setup_entry(
     """Set up switch device."""
     coordinator: LivisiDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
 
-    async def async_discover_device(device) -> None:
+    @callback
+    def handle_coordinator_update() -> None:
         """Add switch."""
-        livisi_switch: SwitchEntity = await create_entity(
-            hass, config_entry, device, coordinator
-        )
-        LOGGER.debug("Include device type: %s", device.get("type"))
-        async_add_entities([livisi_switch])
+        shc_devices: list[dict[str, Any]] = coordinator.data
+        for device in shc_devices:
+            if device["type"] == PSS_DEVICE_TYPE:
+                if device not in coordinator.switch_devices:
+                    livisi_switch: SwitchEntity = create_entity(
+                        config_entry, device, coordinator
+                    )
+                    LOGGER.debug("Include device type: %s", device.get("type"))
+                    coordinator.switch_devices.append(device)
+                    async_add_entities([livisi_switch])
 
-    config_entry.async_on_unload(
-        async_dispatcher_connect(hass, LIVISI_DISCOVERY_NEW, async_discover_device)
-    )
+    coordinator.async_add_listener(handle_coordinator_update)
```

Comment:
We should still use `config_entry.async_on_unload` to unsubscribe the coordinator listener.